### PR TITLE
prov/tcp: Changes in preparation of zero copy support

### PR DESCRIPTION
**Pull request by shefty**
A couple of patches are cleanups which can be pulled out and applied separately.  The rest of the series is an attempt to add support for MSG_ZEROCOPY on larger transfers.  The changes have been tested locally, which forces zerocopy off, so while I believe the general implementation is there, it needs testing where zero copy is supported, plus broad testing on multiple platforms to verify that portability wasn't broken.  And we should probably see if it makes any difference in performance.

More work is needed to make the settings configurable, including forcing zero copy off.  Plus documentation is lacking, which aligns the feature nicely with the zerocopy documentation.

### CODE REVIEW

File: prov/tcp/src/tcpx_msg.c [hunk 1]
```diff
@@ -562,8 +572,10 @@ tcpx_tsenddata(struct fid_ep *fid_ep, const void *buf, size_t len, void *desc,
 			   FI_TAGGED | FI_SEND;
 
 	if (ep->util_ep.tx_op_flags &
-	    (FI_TRANSMIT_COMPLETE | FI_DELIVERY_COMPLETE))
+	    (FI_TRANSMIT_COMPLETE | FI_DELIVERY_COMPLETE)) {
 		tx_entry->hdr.base_hdr.flags |= TCPX_DELIVERY_COMPLETE;
+		tx_entry->flags |= TCPX_NEED_ACK;
+	}
```
Comment (shefty) [inline prov/tcp/src/tcpx_msg.c:578]: See about isolating this into some sort of set_ack_flags function

File: prov/tcp/src/tcpx_progress.c [hunk 2]
```diff
@@ -99,36 +112,48 @@ void tcpx_progress_tx(struct tcpx_ep *ep)
...
+				 * data, so we can skip the async completion.
+				 */
 				slist_insert_tail(&tx_entry->entry,
-						  &ep->tx_rsp_pend_queue);
+						  &ep->need_ack_queue);
+			} else if ((tx_entry->flags & TCPX_ASYNC) &&
+				   (ofi_val32_gt(tx_entry->async_index,
+						 ep->bsock.done_index))) {
+				slist_insert_tail(&tx_entry->entry,
+						  &ep->async_queue);
```
Comment (shefty) [inline prov/tcp/src/tcpx_progress.c:132]: This can be if - else if - else if - else, will outdent code.

File: prov/tcp/src/tcpx_ep.c [hunk 3]
```diff
@@ -63,6 +63,27 @@ void tcpx_hdr_bswap(struct tcpx_base_hdr *hdr)
...
+}
+
+static void tcpx_config_bsock(struct ofi_bsock *bsock)
+{
+	int ret, val = 0;
+	socklen_t len = sizeof(val);
+
+	ret = getsockopt(bsock->sock, SOL_SOCKET, SO_ZEROCOPY, &val, &len);
+	if (!ret && val)
+		bsock->zerocopy_size = OFI_ZEROCOPY_SIZE;
```
Comment (sydidelot) [inline prov/tcp/src/tcpx_ep.c:81]: It seems that we cannot easily figure out if zerocopy is enabled.
It would be great if libfabric could return to the application what specific provider features are enabled for the endpoint (zero-copy, io_uring, etc...)

File: prov/tcp/src/tcpx_progress.c [hunk 4]
```diff
@@ -666,6 +685,24 @@ void tcpx_progress_rx(struct tcpx_ep *ep)
 		tcpx_ep_disable(ep, 0);
 }
 
+void tcpx_progress_err(struct tcpx_ep *ep)
```
Comment (sydidelot) [inline prov/tcp/src/tcpx_progress.c:688]: I understand that ZEROCOPY support reports async completion in an error queue but I find the function name a bit confusing. I would suggest renaming it to something like `tcpx_progress_async`.

File: prov/tcp/src/tcpx_progress.c [hunk 5]
```diff
@@ -666,6 +685,24 @@ void tcpx_progress_rx(struct tcpx_ep *ep)
...
+{
+	struct tcpx_xfer_entry *xfer;
+	uint32_t done;
+
+	done = ofi_bsock_async_done(&tcpx_prov, &ep->bsock);
+	while (!slist_empty(&ep->async_queue)) {
+		xfer = container_of(ep->async_queue.head,
+				    struct tcpx_xfer_entry, entry);
+		if (ofi_val32_gt(xfer->async_index, done))
```
Comment (sydidelot) [inline prov/tcp/src/tcpx_progress.c:697]: That's probably a stupid question but why not calling "xfer->async_index - done > 0" directly? Especially as both variables are uint32_t already.

File: src/common.c [hunk 6]
```diff
@@ -1082,7 +1083,17 @@ ssize_t ofi_bsock_send(struct ofi_bsock *bsock, const void *buf, size_t *len)
 	}
 
 	assert(!ofi_bsock_tosend(bsock));
-	ret = ofi_send_socket(bsock->sock, buf, bytes, MSG_NOSIGNAL);
+	if (bytes > bsock->zerocopy_size) {
+		ret = ofi_send_socket(bsock->sock, buf, bytes,
+				      MSG_NOSIGNAL | OFI_ZEROCOPY);
+		if (ret >= 0) {
+			bsock->async_index++;
+			*len = ret;
```
Comment (sydidelot) [inline src/common.c:1091]: Shouldn't we change the API of ofi_bsock_send() and always return the number of bytes transferred (or waiting for transmission) in the in/out argument `len`, even if zerocopy is not used?

File: src/common.c [hunk 7]
```diff
@@ -1129,7 +1140,17 @@ ssize_t ofi_bsock_sendv(struct ofi_bsock *bsock, const struct iovec *iov,
 	msg.msg_iov = (struct iovec *) iov;
 	msg.msg_iovlen = cnt;
 
-	ret = ofi_sendmsg_tcp(bsock->sock, &msg, MSG_NOSIGNAL);
+	if (bytes > bsock->zerocopy_size) {
+		ret = ofi_sendmsg_tcp(bsock->sock, &msg,
+				      MSG_NOSIGNAL | OFI_ZEROCOPY);
+		if (ret >= 0) {
+			bsock->async_index++;
+			*len = ret;
```
Comment (sydidelot) [inline src/common.c:1148]: Same question as before:  shouldn't we change the API of ofi_bsock_sendv() and always return the number of bytes transferred (or waiting for transmission) in the in/out argument `len`, even if zerocopy is not used?

**Review by sydidelot — COMMENTED** [review]
Sean, here is a couple of comments for the patch.

File: prov/tcp/src/tcpx_ep.c (hunk 3 above)
Comment (shefty) [reply prov/tcp/src/tcpx_ep.c:81]: My intent is to follow this with an option to configure it.  I can pass the provider to this, so we can print the result.

File: prov/tcp/src/tcpx_progress.c (hunk 4 above)
Comment (shefty) [reply prov/tcp/src/tcpx_progress.c:688]: I actually had it named that originally, then changed it to align with the POLL event IN->rx, OUT->tx, ERR->err.  It only handles zerocopy, so I could also go with progress_zerocopy, though I'm considering finding a shorter way of saying zerocopy (zcp? zerocp? bleh).

File: prov/tcp/src/tcpx_progress.c (hunk 5 above)
Comment (shefty) [reply prov/tcp/src/tcpx_progress.c:697]: Not a stupid question at all.  These are 32-bit values that will wrap fairly quickly (probably seconds in a bw test).  If you look at the macro, it casts the results from u32 to int32, which handles the wrapping.  That works as long as index and done never drift farther apart than half the range.

File: src/common.c (hunk 6 above)
Comment (shefty) [reply src/common.c:1091]: Agree - will update as it makes the API easier to use.  I had it this way to avoid the extra write, but the call sites end up having to update anyway.

**Comment by shefty** [comment]
Zero-copy patch was removed.  Other patches updated based on comments.  Those patches should be ready to merge after CI and another round of review.  The zero-copy patch is being updated and will be re-posted.

File: prov/tcp/src/tcpx_ep.c (hunk 3 above)
Comment (shefty) [reply prov/tcp/src/tcpx_ep.c]: Along the lines of your comment, I've had similar request before and have given it some thought.  The most reasonable option is to have the libfabric core handle this task somehow.  For example, capture the fi_info data used to open the fabric, domain, and endpoint.  And define some mechanism for the core to obtain the used value for any fi_param variables.  Anything beyond that becomes difficult and requires explicit opt-in by the providers.
